safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,201 @@
+// limitations under the License.
+
+package csp
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+var randReader = rand.Reader
+
+// nonceSize is the size of the nonces in bytes. According to the CSP3 spec it should
+// be larger than 16 bytes. 20 bytes was picked to be future proof.
+// https://www.w3.org/TR/CSP3/#security-nonces
+const nonceSize = 20
+
+func generateNonce() string {
+	b := make([]byte, nonceSize)
+	_, err := randReader.Read(b)
+	if err != nil {
+		panic(fmt.Errorf("failed to generate entropy using crypto/rand/RandReader: %v", err))
+	}
+	return base64.StdEncoding.EncodeToString(b)
+}
+
+// Policy defines a CSP policy.
+type Policy struct {
+	// serialize serializes this policy for use in a Content-Security-Policy header
+	// or in a Content-Security-Policy-Report-Only header. A nonce will be provided
+	// to serialize which can be used in 'nonce-{random-nonce}' values in directives.
+	serialize func(nonce string) string
+}
+
+type ctxKey struct{}
+
+// Nonce retrieves the nonce from the given context. If there is no nonce stored
+// in the context, an error will be returned.
+func Nonce(ctx context.Context) (string, error) {
+	v := ctx.Value(ctxKey{})
+	if v == nil {
+		return "", errors.New("no nonce in context")
+	}
+	return v.(string), nil
+}
+
+// StrictCSPBuilder can be used to build a strict, nonce-based CSP.
+//
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+type StrictCSPBuilder struct {
+	// StrictDynamic controls whether script-src should contain the 'strict-dynamic'
+	// value.
+	//
+	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#strict-dynamic
+	// for more info.
+	StrictDynamic bool
+	// UnsafeEval controls whether script-src should contain the 'unsafe-eval' value.
+	// If enabled, the eval() JavaScript function is allowed.
+	UnsafeEval bool
+	// BaseURI controls the base-uri directive. If BaseURI is an empty string the
+	// directive will be set to 'none'. The base-uri directive restricts the URLs
+	// which can be used in a document's <base> element.
+	//
+	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/base-uri
+	// for more info.
+	BaseURI string
+	// ReportURI controls the report-uri directive. If ReportUri is empty, no report-uri
+	// directive will be set.
+	ReportURI string
+}
+
+// Build creates a Policy based on the specified options.
+func (s StrictCSPBuilder) Build() Policy {
+	return Policy{
+		serialize: func(nonce string) string {
+			var b strings.Builder
+
+			// object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-{random}'
+			b.WriteString("object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-")
+			b.WriteString(nonce)
+			b.WriteByte('\'')
+
+			if s.StrictDynamic {
+				b.WriteString(" 'strict-dynamic'")
```

Comment:
We need to add fallbacks for browsers that don't support CSP3. If strict-dynamic is specified we need to use ` https: http:` which disables CSP for (basically only) Safari, that didn't implement it.